common/functions/src/scalars/strings/concat.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+use common_datavalues::prelude::*;
+use common_exception::Result;
+
+use crate::scalars::function_factory::FunctionDescription;
+use crate::scalars::function_factory::FunctionFeatures;
+use crate::scalars::Function;
+
+#[derive(Clone)]
+pub struct ConcatFunction {
+    _display_name: String,
+}
+
+impl ConcatFunction {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+        Ok(Box::new(ConcatFunction {
+            _display_name: display_name.to_string(),
+        }))
+    }
+
+    pub fn desc() -> FunctionDescription {
+        FunctionDescription::creator(Box::new(Self::try_create))
+            .features(FunctionFeatures::default().deterministic())
+    }
+}
+
+impl Function for ConcatFunction {
+    fn name(&self) -> &str {
+        "concat"
+    }
+
+    fn return_type(&self, args: &[DataType]) -> Result<DataType> {
+        for arg in args {
+            if arg.is_null() {
+                return Ok(DataType::Null);
+            }
+        }
+        Ok(DataType::String)
+    }
+
+    fn variadic_arguments(&self) -> Option<(usize, usize)> {
+        Some((1, 1024))
+    }
+
+    fn nullable(&self, _input_schema: &DataSchema) -> Result<bool> {
+        Ok(false)
+    }
+
+    fn eval(&self, columns: &DataColumnsWithField, _input_rows: usize) -> Result<DataColumn> {
+        let mut strs: Vec<u8> = Vec::new();
+        for column_field in columns {
+            if column_field.data_type().is_null() {
+                return Ok(DataColumn::Constant(DataValue::Null, 1));
+            }
+            match column_field
+                .column()
+                .cast_with_type(&DataType::String)?
+                .try_get(0)?
```

Comment:
It's just the first element?

Reply:
each column only has one element as part of string, may have many columns to concat

Reply:
Yes, but here is a column vectorized calculation. Everything is columnable.